test/test_orbit/test_orbit.cpp

```diff
@@ -0,0 +1,372 @@
+#include <limits>
+#include <lin.hpp>
+#include <gnc/constants.hpp>
+#include <gnc/config.hpp>
+
+
+//UTILITY MACROS
+#include <unity.h>
+#include "../custom_assertions.hpp"
+
+#define NANOSECONDS_IN_WEEK (7ULL*24ULL*60ULL*60ULL*1'000'000'000ULL)
+
+//0 or 1 that can't be calculated at compile time.
+#ifdef DESKTOP
+#define notcompiletime (rand()%2)
+#else
+#include <Arduino.h>
+#define notcompiletime (millis()%2)
+#endif
+
+
+
+
+
+/**
+ * Recursive Lin Tensor determinant function.
+ * */
+template<typename T, int MR>
+T det(const lin::Matrix<T, 0, 0, MR, MR>& x){
+    int n= x.rows();
+    int m= x.cols();
+    if (m!=n){
+        assert(0);
+    }
+    assert(n>0);
+    if (n==1){
+        return x(0,0);
+    }
+    if (n==2){
+        return x(0,0)*x(1,1)-x(1,0)*x(0,1);
+    }
+    T result=0;
+
+    for (int i=0; i<n; i++){
+        //x without row 0 or col i
+        lin::Matrix<T,0,0,MR,MR> submat(n-1,n-1);
+        if (i==0){
+            submat= lin::ref<0, 0, MR, MR> (x, 1, i+1, n-1, n-1-i);
+        }else if (i==n-1){
+            submat= lin::ref<0, 0, MR, MR> (x, 1, 0, n-1, i);
+        }else {
+            lin::ref<0, 0, MR, MR> (submat, 0, 0, n-1, i) = lin::ref<0, 0, MR, MR> (x, 1, 0, n-1, i);
+            lin::ref<0, 0, MR, MR> (submat, 0, i, n-1, n-1-i) = lin::ref<0, 0, MR, MR> (x, 1, i+1, n-1, n-1-i);
+        }
+        result+= (i%2?-1:1)*x(0,i)*det<T,MR>(submat);
+    }
+    return result;
+}
+
+//grace orbit initial
+const orb::Orbit gracestart(uint64_t(gnc::constant::init_gps_week_number)*NANOSECONDS_IN_WEEK,{-6522019.833240811L, 2067829.846415895L, 776905.9724453629L},{941.0211143841228L, 85.66662333729801L, 7552.870253470936L});
+
+//grace orbit 100 seconds later
+const orb::Orbit grace100s(uint64_t(gnc::constant::init_gps_week_number)*NANOSECONDS_IN_WEEK+100'000'000'000ULL,{-6388456.55330517L, 2062929.296577276L, 1525892.564091281L},{1726.923087560988L, -185.5049475128178L, 7411.544615026139L});
+
+//grace orbit 13700 seconds later
+const orb::Orbit grace13700s(uint64_t(gnc::constant::init_gps_week_number)*NANOSECONDS_IN_WEEK+13700'000'000'000ULL,{1579190.147083268L, -6066459.613667888L, 2785708.976728437L},{480.8261476296949L, -3083.977113497177L, -6969.470748202005L});
+
+//earth rate in ecef (rad/s)
+lin::Vector3d earth_rate_ecef= {0.000000707063506E-4,-0.000001060595259E-4,0.729211585530000E-4};
+
+void test_basic_constructors() {
+    orb::Orbit x;
+    TEST_ASSERT_FALSE(x.valid());
+
+    //test valid orbit
+    TEST_ASSERT_TRUE(gracestart.valid());
+    x=gracestart;
+    TEST_ASSERT_TRUE(x.valid());
+
+    //test invalid orbit
+    lin::Vector3d r2 {0.0L, 0.0L, 0.0L};
+    lin::Vector3d v2 {941.0211143841228L, 85.66662333729801L, 7552.870253470936L};
+    uint64_t t2= uint64_t(gnc::constant::init_gps_week_number)*NANOSECONDS_IN_WEEK;
+    orb::Orbit y2(t2,r2,v2);
+    TEST_ASSERT_FALSE(y2.valid());
+
+    //test NAN invalid orbit
+    lin::Vector3d r3 {0.0, std::numeric_limits<double>::quiet_NaN(), 0.0};
```

Comment:
Just an FYI, there are two constants for NaN values: `constant::nan` for doubles and `constant::nan_f` for floats. Using those would be a little cleaner.